flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -222,19 +226,54 @@ public SingleInputGate(
 	}
 
 	@Override
-	public void setup() throws IOException, InterruptedException {
+	public void setup() throws IOException {
 		checkState(this.bufferPool == null, "Bug in input gate setup logic: Already registered buffer pool.");
 		// assign exclusive buffers to input channels directly and use the rest for floating buffers
 		assignExclusiveSegments();
 
 		BufferPool bufferPool = bufferPoolFactory.get();
 		setBufferPool(bufferPool);
+	}
+
+	@Override
+	public void readRecoveredState(ExecutorService executor, ChannelStateReader reader) throws IOException {
+		internalRequestPartitions(() -> executor.submit(() -> {
+
+			for (InputChannel inputChannel : inputChannels.values()) {
+				if (inputChannel instanceof RemoteInputChannel) {
+					try {
+						((RemoteInputChannel) inputChannel).readRecoveredState(reader);
+					} catch (Throwable t) {
+						inputChannel.setError(t);
+						return;
+					}
+				}
```

Comment:
I think this block of code is forgetting about `getUnconsumedBuffer()` from the `LocalInputChannel` that was spilled as part of the input data.